File: src/test/java/auctionsniper/FakeAuctionServer.java
package auctionsniper;

import org.hamcrest.Matcher;
import org.jivesoftware.smack.ConnectionConfiguration;
import org.jivesoftware.smack.SmackException;
import org.jivesoftware.smack.XMPPException;
import org.jivesoftware.smack.chat.Chat;
import org.jivesoftware.smack.chat.ChatManager;
import org.jivesoftware.smack.tcp.XMPPTCPConnection;
import org.jivesoftware.smack.tcp.XMPPTCPConnectionConfiguration;
import org.jxmpp.jid.impl.JidCreate;
import org.jxmpp.jid.parts.Resourcepart;
import org.jxmpp.stringprep.XmppStringprepException;

import java.io.IOException;

import static auctionsniper.AppConstants.*;
import static java.lang.String.format;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

public class FakeAuctionServer {

    private static final String AUCTION_PASSWORD = "auction";

    private final SingleMessageListener messageListener = new SingleMessageListener();
    private final String itemId;
    private final XMPPTCPConnection connection;

    private Chat currentChat;

    public FakeAuctionServer(String hostname, String itemId) throws XmppStringprepException {
        this.itemId = itemId;
        XMPPTCPConnectionConfiguration config = XMPPTCPConnectionConfiguration.builder()
                .setHost(hostname)
                .setXmppDomain(JidCreate.from(XMPP_DOMAIN).asDomainBareJid())
                .setSecurityMode(ConnectionConfiguration.SecurityMode.disabled)
//                .setDebuggerEnabled(true)
                .build();
        this.connection = new XMPPTCPConnection(config);
    }

    public void startSellingItem() throws IOException, XMPPException, SmackException, InterruptedException {
        connection.connect();
        connection.login(format(ITEM_ID_AS_LOGIN, itemId), AUCTION_PASSWORD, Resourcepart.from(AUCTION_RESOURCE));
        ChatManager.getInstanceFor(connection).addChatListener(
                (chat, createdLocally) -> {
                    currentChat = chat;
                    chat.addMessageListener(messageListener);
                });
    }

    public void hasReceivedJoinRequestFrom(String sniperId) throws InterruptedException {
        receivesAMessageMatching(sniperId, equalTo(JOIN_COMMAND_FORMAT));
    }

    private void receivesAMessageMatching(String sniperId, Matcher<? super String> messageMatcher) throws InterruptedException {
        messageListener.receivesAMessage(messageMatcher);
        assertThat(currentChat.getParticipant(), equalTo(sniperId));
    }

    public void reportPrice(int price, int increment, String bidder) throws SmackException, InterruptedException {
        currentChat.sendMessage(format(PRICE_COMMAND_FORMAT, price, increment, bidder));
    }

    public void hasReceivedBid(int bid, String sniperId) throws InterruptedException {
        receivesAMessageMatching(sniperId, equalTo(format(BID_COMMAND_FORMAT, bid)));
    }

    public void announceClosed() throws SmackException, InterruptedException {
        currentChat.sendMessage(CLOSE_COMMAND_FORMAT);
    }

    public void stop() {
        connection.disconnect();
    }

    public String getItemId() {
        return itemId;
    }
}

File: src/endToEndTest/kotlin/test/auctionsniper/support/FakeAuctionServer.kt
package test.auctionsniper.support

import auctionsniper.xmpp.XMPPAuction
import auctionsniper.xmpp.XMPPAuctionHouse
import org.hamcrest.CoreMatchers
import org.hamcrest.Matcher
import org.hamcrest.MatcherAssert
import org.jivesoftware.smack.ConnectionConfiguration
import org.jivesoftware.smack.chat.Chat
import org.jivesoftware.smack.chat.ChatManager
import org.jivesoftware.smack.tcp.XMPPTCPConnection
import org.jivesoftware.smack.tcp.XMPPTCPConnectionConfiguration
import org.jxmpp.jid.impl.JidCreate
import org.jxmpp.jid.parts.Resourcepart
import test.auctionsniper.support.Constants.AUCTION_PASSWORD
import test.auctionsniper.xmpp.AuctionMessageTranslatorTest

class FakeAuctionServer(hostname: String) {

    private val messageListener = SingleMessageListener()
    private val connection: XMPPTCPConnection
    private lateinit var itemId: String
    private lateinit var currentChat: Chat

    init {
        val config = XMPPTCPConnectionConfiguration.builder()
            .setHost(hostname)
            .setXmppDomain(JidCreate.from(XMPPAuctionHouse.XMPP_DOMAIN).asDomainBareJid())
            .setSecurityMode(ConnectionConfiguration.SecurityMode.disabled)
//          .setDebuggerEnabled(true)
            .build()
        connection = XMPPTCPConnection(config)
    }

    fun stop() {
        connection.disconnect()
    }

    fun startSellingItem(itemId: String) {
        this.itemId = itemId

        connection.connect()
        connection.login(
            java.lang.String.format(XMPPAuction.ITEM_ID_AS_LOGIN, itemId),
            AUCTION_PASSWORD,
            Resourcepart.from(XMPPAuctionHouse.AUCTION_RESOURCE)
        )
        val listener = { chat: Chat, _: Boolean ->
            currentChat = chat
            chat.addMessageListener(messageListener)
        }
        ChatManager.getInstanceFor(connection).addChatListener(listener)
    }

    fun hasReceivedJoinRequestFrom(sniperId: String) {
        receivesAMessageMatching(sniperId, CoreMatchers.equalTo(XMPPAuction.JOIN_COMMAND_FORMAT))
    }

    fun hasReceivedBid(bid: Int, sniperId: String) {
        receivesAMessageMatching(
            sniperId,
            CoreMatchers.equalTo(java.lang.String.format(XMPPAuction.BID_COMMAND_FORMAT, bid))
        )
    }

    fun announceClosed() {
        currentChat.sendMessage(AuctionMessageTranslatorTest.CLOSE_COMMAND_FORMAT)
    }

    fun reportPrice(price: Int, increment: Int, bidder: String) {
        currentChat.sendMessage(
            java.lang.String.format(
                AuctionMessageTranslatorTest.PRICE_COMMAND_FORMAT,
                price,
                increment,
                bidder
            )
        )
    }

    fun itemId(): String = itemId

    fun sendInvalidMessageContaining(message: String) {
        currentChat.sendMessage(message)
    }

    private fun receivesAMessageMatching(sniperId: String, messageMatcher: Matcher<in String>) {
        messageListener.receivesAMessage(messageMatcher)
        MatcherAssert.assertThat(currentChat.participant.asUnescapedString(), CoreMatchers.equalTo(sniperId))
    }
}
